### PR TITLE
feat(nextjs): Clock skew error message

### PR DESCRIPTION
**Pull request by BRKalow**
## Type of change

- [ ] 🐛 Bug fix
- [ ] 🌟 New feature
- [ ] 🔨 Breaking change
- [x] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

## Packages affected

- [ ] `@clerk/clerk-js`
- [ ] `@clerk/clerk-react`
- [x] `@clerk/nextjs`
- [ ] `@clerk/remix`
- [ ] `@clerk/types`
- [ ] `@clerk/themes`
- [ ] `@clerk/localizations`
- [ ] `@clerk/clerk-expo`
- [x] `@clerk/backend`
- [ ] `@clerk/clerk-sdk-node`
- [ ] `@clerk/shared`
- [ ] `@clerk/fastify`
- [ ] `@clerk/chrome-extension`
- [ ] `gatsby-plugin-clerk`
- [ ] `build/tooling/chore`

## Description

- [ ] `npm test` runs as expected.
- [ ] `npm run build` runs as expected.

Attempt to detect clock skew in development based on the `requestState`. If detected, we log a helpful error message.

Example output:
```
Error: Clerk: Clock skew detected. This usually means that your server's system clock is inaccurate. Clerk will continuously try to issue new tokens, as the existing ones will be treated as "expired" due to clock skew.

To resolve this issue, make sure your system's clock is set to the correct time (e.g. turn off and on automatic time synchronization).

---

JWT cannot be used prior to not before date claim (nbf). Not before date: Wed Aug 30 2023 13:18:40 GMT-0500 (Central Daylight Time); Current date: Wed Aug 30 2023 13:17:56 GMT-0500 (Central Daylight Time); (reason=token-not-active-yet, token-carrier=cookie)
```

related to JS-675

### CODE REVIEW

**Review by nikosdouvlis — APPROVED** [review]
This makes sense to me!

File: packages/nextjs/src/server/errors.ts [hunk 1]
```diff
@@ -34,20 +34,24 @@ export const getAuthAuthHeaderMissing = () =>
 export const authAuthHeaderMissing = () =>
   "Clerk: auth() was called but it looks like you aren't using `authMiddleware` in your middleware file. Please use `authMiddleware` and make sure your middleware matcher is configured correctly and it matches this route or page. See https://clerk.com/docs/quickstarts/get-started-with-nextjs";
 
+export const clockSkewDetected = (verifyMessage: string) =>
+  `Clerk: Clock skew detected. This usually means that your server's system clock is inaccurate. Clerk will continuously try to issue new tokens, as the existing ones will be treated as "expired" due to clock skew.
```
[the remark this answers is not in the document]
Comment (BRKalow) [reply packages/nextjs/src/server/errors.ts:38]: Yep, should only be in development. Great call. I'll tweak the wording here.

File: packages/nextjs/src/server/authMiddleware.ts [hunk 2]
```diff
@@ -334,20 +342,41 @@ const isRequestMethodIndicatingApiRoute = (req: NextRequest): boolean => {
...
 
+/**
+ * In development, attempt to detect clock skew based on the requestState. This check should run when requestState.isInterstitial is true. If detected, we throw an error.
+ */
+const assertClockSkew = (requestState: RequestState, opts: AuthMiddlewareParams): void => {
+  if (!isDevelopmentFromApiKey(opts.secretKey || SECRET_KEY)) {
+    return;
+  }
+
+  if (requestState.reason === 'token-not-active-yet') {
```
[the remark this answers is not in the document]
Comment (BRKalow) [reply packages/nextjs/src/server/authMiddleware.ts:353]: @nikosdouvlis I would have to import from the `dist`, what's our stance on that?

```ts
import { TokenVerificationErrorReason } from '@clerk/backend/dist/types/tokens/errors';
```

**Comment by BRKalow** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @BRKalow - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/backend | 0.28.1-snapshot.b493055 |
| @clerk/chrome-extension | 0.3.29-snapshot.b493055 |
| @clerk/clerk-js | 4.56.2-snapshot.b493055 |
| eslint-config-custom | 0.3.0 |
| @clerk/clerk-expo | 0.18.20-snapshot.b493055 |
| @clerk/fastify | 0.6.5-snapshot.b493055 |
| gatsby-plugin-clerk | 4.4.6-snapshot.b493055 |
| @clerk/localizations | 1.25.0 |
| @clerk/nextjs | 4.23.4-snapshot.b493055 |
| @clerk/clerk-react | 4.24.0 |
| @clerk/remix | 2.10.1-snapshot.b493055 |
| @clerk/clerk-sdk-node | 4.12.4-snapshot.b493055 |
| @clerk/shared | 0.22.0 |
| @clerk/themes | 1.7.5 |
| @clerk/types | 3.50.0 |

Tip: use the snippet copy button below to quickly install the required packages.
```sh
# @clerk/backend
npm i @clerk/backend@0.28.1-snapshot.b493055
```
```sh
# @clerk/chrome-extension
npm i @clerk/chrome-extension@0.3.29-snapshot.b493055
```
```sh
# @clerk/clerk-js
npm i @clerk/clerk-js@4.56.2-snapshot.b493055
```
```sh
# eslint-config-custom
npm i eslint-config-custom@0.3.0
```
```sh
# @clerk/clerk-expo
npm i @clerk/clerk-expo@0.18.20-snapshot.b493055
```
```sh
# @clerk/fastify
npm i @clerk/fastify@0.6.5-snapshot.b493055
```
```sh
# gatsby-plugin-clerk
npm i gatsby-plugin-clerk@4.4.6-snapshot.b493055
```
```sh
# @clerk/localizations
npm i @clerk/localizations@1.25.0
```
```sh
# @clerk/nextjs
npm i @clerk/nextjs@4.23.4-snapshot.b493055
```
```sh
# @clerk/clerk-react
npm i @clerk/clerk-react@4.24.0
```
```sh
# @clerk/remix
npm i @clerk/remix@2.10.1-snapshot.b493055
```
```sh
# @clerk/clerk-sdk-node
npm i @clerk/clerk-sdk-node@4.12.4-snapshot.b493055
```
```sh
# @clerk/shared
npm i @clerk/shared@0.22.0
```
```sh
# @clerk/themes
npm i @clerk/themes@1.7.5
```
```sh
# @clerk/types
npm i @clerk/types@3.50.0
```

File: packages/nextjs/src/server/authMiddleware.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (dimkl) [reply packages/nextjs/src/server/authMiddleware.ts:353]: Why not use the `requestState.action: TokenVerificationErrorAction.EnsureClockSync` ?

**Review by dimkl — APPROVED** [review]

**Comment by clerk-cookie** [comment]
This PR has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.

**Pull request merged**